cpp/src/lib.rs

```diff
@@ -92,12 +125,58 @@ extern crate cpp_macros;
 #[doc(hidden)]
 pub use cpp_macros::*;
 
+/// Internal macro which is used to locate the rust! invocations in the
+/// C++ code embbeded in cpp! invocation, to translate them into extern
+/// functions
+#[doc(hidden)]
+#[macro_export]
+macro_rules! __cpp_internal {
+    (@find_rust_macro rust!($($rust_body:tt)*) $($rest:tt)*) => {
+        __cpp_internal!{ @expand_rust_macro $($rust_body)* }
+        __cpp_internal!{ @find_rust_macro $($rest)* }
+    };
+    (@find_rust_macro ( $($in:tt)* ) $($rest:tt)* ) =>
+        { __cpp_internal!{ @find_rust_macro $($in)* $($rest)* }  };
+    (@find_rust_macro [ $($in:tt)* ] $($rest:tt)* ) =>
+        { __cpp_internal!{ @find_rust_macro $($in)* $($rest)* }  };
+    (@find_rust_macro { $($in:tt)* } $($rest:tt)* ) =>
+        { __cpp_internal!{ @find_rust_macro $($in)* $($rest)* }  };
+    (@find_rust_macro $t:tt $($rest:tt)*) =>
+        { __cpp_internal!{ @find_rust_macro $($rest)* } };
+    (@find_rust_macro) => {};
+
+    (@expand_rust_macro $i:ident [$($an:ident : $at:ty as $ac:tt),*] {$($body:tt)*}) => {
+        #[no_mangle]
+        #[doc(hidden)]
+        pub extern "C" fn $i($($an : *const $at),*) {
+            $(let $an : $at = unsafe { std::ptr::read($an) };)*
+            $($body)*
+        }
+    };
+    (@expand_rust_macro $i:ident [$($an:ident : $at:ty as $ac:tt),*] -> $rt:ty as $rc:tt {$($body:tt)*}) => {
+        #[no_mangle]
+        #[doc(hidden)]
+        pub extern "C" fn $i($($an : *const $at, )* rt : *mut $rt) -> *mut $rt {
+            $(let $an : $at = unsafe { std::ptr::read($an) };)*
+            let lambda = || {$($body)*};
+            unsafe { std::ptr::write(rt, lambda()) };
```

Comment:
We probably want to panic::catch_unwind here, so that we can avoid unwinding across language boundaries.